Rewrite products of exponents as exponent of sum

tamastokes opened the pull request:
Hi,

Could someone please check whether I'm on the right track with this attempt to solve https://github.com/pymc-devs/pytensor/issues/54

This commit adds rewrites for e^x * e^y as e^(x+y) and e^x / e^y as e^(x-y). If you confirm that this is more or less the appropriate way, I'm happy to continue with power(base, x) * power(base, y) => power(base, x + y)

Apart from the general request for comments, I have a specific question too: I don't yet really understand how exactly the added decorators (@register_canonicalize and @register_specialize) change behaviour. I'd naively presume this particular node rewriter should be a genuine specializer, not sure about the canonicalize part though. Is there any document somewhere that explains this? Unfortunately, I haven't come across anything related to this so far.

Also, what else do you think would be useful to test here?

Thanks,
Tamás

ricardoV94 commented on pytensor/tensor/rewriting/math.py:
```diff
@@ -423,6 +423,35 @@ def local_sumsqr2dot(fgraph, node):
 
+@register_canonicalize
+@register_specialize
+@node_rewriter([Elemwise])
```
This will make it so the rewrite is only called on these nodes, and you can skip the first `isinstance` check

```suggestion
@node_rewriter([mul, true_div])
```

ricardoV94 commented on pytensor/tensor/rewriting/math.py:
```diff
@@ -423,6 +423,35 @@ def local_sumsqr2dot(fgraph, node):
+    if (
+        isinstance(node.op, Elemwise)
+        and isinstance(node.op.scalar_op, (aes.Mul, aes.TrueDiv))
+        and node.inputs[0].owner
+        and isinstance(node.inputs[0].owner.op, Elemwise)
+        and isinstance(node.inputs[0].owner.op.scalar_op, aes.Exp)
+        and node.inputs[1].owner
+        and isinstance(node.inputs[1].owner.op, Elemwise)
+        and isinstance(node.inputs[1].owner.op.scalar_op, aes.Exp)
+    ):
```
`mul` can actually have arbitrary number of inputs such as in `mul(5, 6, 7)`, so the rewrite should be aware of that.

ricardoV94 replied:
The TrueDiv only accepts two inputs, never more, so you can keep treating as before.

You can assume multiplication is commutative yes. Careful that you don't want to sum (which reduces dimensions) but add all the cases that can be merged. Like `mul`, `add` allows arbitrary number of inputs.

ricardoV94 reviewed (changes requested):
This looks great. I left some comments below

ricardoV94 commented:
> Apart from the general request for comments, I have a specific question too: I don't yet really understand how exactly the added decorators (@register_canonicalize and @register_specialize) change behaviour. 

The decorators convert the function into a "proper rewriter". Something you can call `.transform` with.

```python
import pytensor
import pytensor.tensor as pt
from pytensor.tensor.rewriting.math import local_log1p

x = pt.scalar("x")
y = pt.log(1 + x)
new_y = local_log1p.transform(None, y.owner)
pytensor.dprint(new_y)
```

> I'd naively presume this particular node rewriter should be a genuine specializer, not sure about the canonicalize part though. Is there any document somewhere that explains this? Unfortunately, I haven't come across anything related to this so far.

I think it works fine just as a specialization. This canonicalize database is supposed to include rewrites that convert wildely different yet equivalent graphs into a common or "canonical" form, so that other rewrites can build on top of these forms. The specialize database is supposed to include rewrites that convert a graph into more efficient forms. It's not always obvious which category a rewrite should go into.

ricardoV94 commented:
I realize now you were asking only about the canonicalize and specialize decorators. Those simply register the rewrite in the respective databases.

tamastokes commented:
Hi, I've implemented what we discussed above re the multiplication of more than two factors, and I added a similar node rewriter for the pow() op too. Please review. Thanks!

ricardoV94 commented on pytensor/tensor/rewriting/math.py:
```diff
@@ -423,6 +423,108 @@ def local_sumsqr2dot(fgraph, node):
 
+@register_specialize
+@node_rewriter([mul, true_div])
+def local_mulexp2expadd(fgraph, node):
```
nitpick, the other rewrites in the codebase follow a `word_word` and `_to_` convention, so this would be `local_mul_exp_to_exp_add`.

ricardoV94 commented on pytensor/tensor/rewriting/math.py:
```diff
@@ -423,6 +423,108 @@ def local_sumsqr2dot(fgraph, node):
+    This rewrite detects e^x * e^y and converts it to e^(x+y).
+    Similarly, e^x / e^y becomes e^(x-y).
+    """
+    if isinstance(node.op, Elemwise) and isinstance(
+        node.op.scalar_op, (aes.Mul, aes.TrueDiv)
+    ):
```
You don't need this first `isinstance` check, the rewrite will only be called to these `Op`s. Other rewrites in the codebase always check at the top, because they predated those changes

tamastokes replied:
Oh, gotcha, this is because mul is actually a shorthand for Elemwise{mul,...}, right? Before recognising this I had thought the caller calls this function for any node that looks like */mul/* so I have to do the isinstance(Elemwise) check, but now I get that because of mul's decorator to wrap it into an Elemwise, the the caller only calls this function for nodes that look like */Elemwise/mul/*. At least I hope I now understand it correctly. :)

I made all the changes you requested, so I hope we're good now. Thanks!

ricardoV94 replied:
Yes, you're correct

ricardoV94 commented on pytensor/tensor/rewriting/math.py:
```diff
@@ -423,6 +423,108 @@ def local_sumsqr2dot(fgraph, node):
+    if isinstance(node.op, Elemwise) and isinstance(
+        node.op.scalar_op, (aes.Mul, aes.TrueDiv)
+    ):
+        from collections import defaultdict
```
Since this rewrite will be called almost every time (every time there's a mul or div), it's fine to make the import global

ricardoV94 reviewed (commented):
This looks great, left just some small nitpicks in my review

ricardoV94 commented on pytensor/tensor/rewriting/math.py:
```diff
@@ -423,6 +424,102 @@ def local_sumsqr2dot(fgraph, node):
+    This rewrite detects e^x * e^y and converts it to e^(x+y).
+    Similarly, e^x / e^y becomes e^(x-y).
+    """
+    if isinstance(node.op.scalar_op, (aes.Mul, aes.TrueDiv)):
```
Not even this isinstance is needed

tamastokes replied:
You're right, fixed.

ricardoV94 commented:
I rebased your PR so that the tests run. They were failing due to #195

ricardoV94 commented:
The failing tests are running with `FAST_COMPILE` flag set to True, which does not introduce Composite Ops.

You can reproduce by setting
```python
import pytensor
pytensor.config.mode = ”FAST_COMPILE”
```
At the top of the test file.

Other tests usually tweak the mode do include/exclude fusion rewrites, so you can do the same.

ricardoV94 commented on pytensor/tensor/rewriting/math.py:
```diff
@@ -423,6 +424,100 @@ def local_sumsqr2dot(fgraph, node):
                     return [new_out]
 
 
+@register_specialize
```
Rethinking this, it may make sense to register these new rewrites in canonicalize as well.

tamastokes replied:
Hi, after looking around how it's resolved elsewhere and some experimentation with different options, I've ended up adding the following line at the beginning of the test function, which seems to have resolved the issue on my local runs in both with and without the FAST_COMPILE flag.

`    mode = get_mode("FAST_RUN") if config.mode == "FAST_COMPILE" else get_default_mode()
`

(Obviously, with passing the mode parameter to all function() calls)

If you're ok with this, on which branch do you want me to commit this change? The original one or the the one you had rebased that branch earlier?

Also, do you want me to add back the @register_canonicalize decorators?

Thanks!

ricardoV94 replied:
I think a more clean solution is to use `mode=get_default_mode.excluding("fusion")`

Then the tests will also be more straightforward because you don't need to check the graph inside the Composites

And yes, let's try to add the `register_canonicalize` to these rewrites and see if it doesn't break anything.

tamastokes replied:
An interesting side effect of making the rewrites canonical is that one of the test case in test_math.py::TestSigmoidRewrites::test_local_sigm_times_exp breaks, as now my rewrites takes precedence over the sigma_times_exp rewrites. More specifically,
`sigma(x) * sigma(-y) * (-exp(-x)) * exp(xy) * exp(y)`
is supposed to become
`(-1) * sigma(-x) * sigma(y) * exp(xy)`
but my rewrites contracts the exp-s before the sigma*exp rewrite could take action, and the result becomes
`(-1) * exp(xy+y-x) * sigma(x) * sigma(-y)`

So, I guess, this is a trade-off, how should we resolve this? (Having run all tests, this is the only one the canonicalisation broke.)

ricardoV94 replied:
We could maybe extend the sigma_exp rewrite to check if the balancing term is inside the exponentiation?

`exp(y - x) * sigmoid(x) -> exp(y) * sigmoid(x)`

Does that make sense? Might have other ramifications I am not seeing.

ricardoV94 replied:
That's a bit challenging looking at the rewrite so let's leave as is (don't canonicalize) and we can open an issue to investigate if it's worth it.

tamastokes replied:
I'm wondering how frequent / marginal such use cases are. Are there any kind of statistics available on how often real world use cases trigger which rewrites?

Meanwhile, I've just committed the changes for the tests, hope they'll be fine this time.

Thanks!

ricardoV94 replied:
Unfortunately, we have no telemetrics to assess types of graphs users create :D

michaelosthege reviewed (approved):
I suppose this can be merged now that all the tests are green?

ricardoV94 commented:
Thanks @tamastokes, great work 👍
